File: src/main/c/cext/tracepoint.c
/*
 * Copyright (c) 2020, 2021 Oracle and/or its affiliates. All rights reserved. This
 * code is released under a tri EPL/GPL/LGPL license. You can use it,
 * redistribute it and/or modify it under the terms of the:
 *
 * Eclipse Public License version 2.0, or
 * GNU General Public License version 2, or
 * GNU Lesser General Public License version 2.1.
 */
#include <truffleruby-impl.h>
#include <ruby/debug.h>

// TracePoint, rb_tracepoint_*

VALUE rb_tracepoint_new(VALUE target_thval, rb_event_flag_t events, void (*func)(VALUE, void *), void *data) {
  // target_thval parameter is unused as of MRI 2.6
  return rb_tr_wrap(polyglot_invoke(RUBY_CEXT, "rb_tracepoint_new", events, func, data));
}

VALUE rb_tracepoint_enable(VALUE tpval) {
  return RUBY_INVOKE(tpval, "enable");
}

VALUE rb_tracepoint_disable(VALUE tpval) {
  return RUBY_INVOKE(tpval, "disable");
}

VALUE rb_tracepoint_enabled_p(VALUE tpval) {
  return RUBY_INVOKE(tpval, "enabled?");
}

File: src/main/ruby/truffleruby/core/tracepoint.rb
# frozen_string_literal: true

# Copyright (c) 2019, 2025 Oracle and/or its affiliates. All rights reserved. This
# code is released under a tri EPL/GPL/LGPL license. You can use it,
# redistribute it and/or modify it under the terms of the:
#
# Eclipse Public License version 2.0, or
# GNU General Public License version 2, or
# GNU Lesser General Public License version 2.1.

class TracePoint
  def self.trace(*events, &handler)
    TracePoint.new(*events, &handler).tap(&:enable)
  end

  def initialize(*events, &handler)
    events = [:line] if events.empty?
    events = events.map { |event| Truffle::Type.coerce_to event, Symbol, :to_sym }
    events.each do |event|
      case event
      when :line, :class, :never
        # Supported event, keep it and pass it below
      else
        raise ArgumentError, "unknown event: #{event}"
      end
    end

    raise ArgumentError, 'must be called with a block' unless handler

    Primitive.tracepoint_initialize self, events.uniq, handler
  end

  def inspect
    if Primitive.trace_point_inside_proc?
      return "#<TracePoint:#{event} #{path}:#{lineno}>"
    end

    state = enabled? ? 'enabled' : 'disabled'
    "#<TracePoint:#{state}>"
  end
end
